esrally/utils/net.py

```diff
@@ -65,39 +65,84 @@ def finish(self):
         self.p.finish()
 
 
+# This function is not meant to be called externally and only exists for unit tests
+def _download_from_s3_bucket(bucket_name, bucket_path, local_path, expected_size_in_bytes=None, progress_indicator=None):
+    # lazily initialize S3 support - we might not need it
+    import boto3.s3.transfer
+
+    class S3ProgressAdapter:
+        def __init__(self, size, progress):
+            self._expected_size_in_bytes = size
+            self._progress = progress
+            self._bytes_read = 0
+
+        def __call__(self, bytes_amount):
+            self._bytes_read += bytes_amount
+            self._progress(self._bytes_read, self._expected_size_in_bytes)
+
+    s3 = boto3.resource("s3")
+    progress_callback = S3ProgressAdapter(expected_size_in_bytes, progress_indicator) if progress_indicator else None
+    s3.Bucket(bucket_name).download_file(bucket_path, local_path,
+                                         Callback=progress_callback,
+                                         Config=boto3.s3.transfer.TransferConfig(use_threads=False))
+
+
+def download_s3(url, local_path, expected_size_in_bytes=None, progress_indicator=None):
+    logger = logging.getLogger(__name__)
+
+    bucket_and_path = url[5:]
+    bucket_end_index = bucket_and_path.find("/")
+    bucket = bucket_and_path[:bucket_end_index]
+    # we need to remove the leading "/"
+    bucket_path = bucket_and_path[bucket_end_index + 1:]
+
+    logger.info("Downloading from S3 bucket [%s] and path [%s] to [%s].", bucket, bucket_path, local_path)
+    _download_from_s3_bucket(bucket, bucket_path, local_path, expected_size_in_bytes, progress_indicator)
+
+    return expected_size_in_bytes
+
+
+def download_http(url, local_path, expected_size_in_bytes=None, progress_indicator=None):
```

Comment:
does it also make sense to add a unit test for `download_http`? I couldn't find if we have a test for download().. although it is a well executed code path so not sure if we need it.

Reply:
Tbh, I attempted to do it but I ended up mocking so much code that I had the impression it's not really worth it.